tensorflow_quantum/python/differentiators/parameter_shift.py

```diff
@@ -102,57 +157,20 @@ def differentiate_analytic(self, programs, symbol_names, symbol_values,
             Backward gradient values for each program & each pauli sum. It has
             the shape of [batch_size, n_symbols].
         """
-
-        # these get used a lot
+        n_ops = tf.gather(tf.shape(pauli_sums), 1)
         n_symbols = tf.gather(tf.shape(symbol_names), 0)
         n_programs = tf.gather(tf.shape(programs), 0)
-        n_ops = tf.gather(tf.shape(pauli_sums), 1)
         # Assume cirq.decompose() generates gates with at most two distinct
         # eigenvalues, which results in two parameter shifts.
         n_shifts = 2
-
-        # STEP 1: Generate required inputs for executor
-        # Deserialize programs and parse the whole parameterized gates
-        # new_programs has [n_symbols, n_param_gates, n_shifts, n_programs].
-        # These new_programs has programs that parameter-shift rule is applied,
-        # so those programs has
-        (new_programs, weights, shifts,
-         n_param_gates) = parameter_shift_util.parse_programs(
-             programs, symbol_names, symbol_values, n_symbols)
-
-        # Reshape & transpose new_programs, weights and shifts to fit into
-        # the input format of tensorflow_quantum simulator.
-        # [n_symbols, n_param_gates, n_shifts, n_programs]
-        new_programs = tf.transpose(new_programs, [0, 2, 3, 1])
-        weights = tf.transpose(weights, [0, 2, 3, 1])
-        shifts = tf.transpose(shifts, [0, 2, 3, 1])
-
-        # reshape everything to fit into expectation op correctly
-        total_programs = n_programs * n_shifts * n_param_gates * n_symbols
-        # tile up and then reshape to order programs correctly
-        flat_programs = tf.reshape(new_programs, [total_programs])
-        flat_shifts = tf.reshape(shifts, [total_programs])
-
-        # tile up and then reshape to order ops correctly
+        (flat_programs, new_symbol_names, flat_weights, flat_perturbations,
+         n_param_gates) = self.get_intermediate_logic(programs, symbol_names,
+                                                      symbol_values, pauli_sums)
+        total_programs = n_param_gates * n_programs * n_shifts * n_symbols
         n_tile = n_shifts * n_param_gates * n_symbols
-        flat_perturbations = tf.concat([
-            tf.reshape(
-                tf.tile(tf.expand_dims(symbol_values, 0),
-                        tf.stack([n_tile, 1, 1])), [total_programs, n_symbols]),
-            tf.expand_dims(flat_shifts, axis=1)
-        ],
-                                       axis=1)
         flat_ops = tf.reshape(
```

Comment:
This isn't being returned by `get_intermediate_logic` function even though it is essential for the expectation calculation below, we shouldn't force a user to have to think harder than they need to on how to call into the op in their own way should they want to.

Reply:
Refactored